src/extension/tools/vscode-node/toolsService.ts

```diff
@@ -91,42 +92,54 @@ export class ToolsService extends BaseToolsService {
 		throw new Error('This method for tests only');
 	}
 
-	getEnabledTools(request: vscode.ChatRequest, filter?: (tool: vscode.LanguageModelToolInformation) => boolean | undefined): vscode.LanguageModelToolInformation[] {
+	getEnabledTools(request: vscode.ChatRequest, endpoint: IChatEndpoint, filter?: (tool: vscode.LanguageModelToolInformation) => boolean | undefined): vscode.LanguageModelToolInformation[] {
 		const toolMap = new Map(this.tools.map(t => [t.name, t]));
 
-		return this.tools.filter(tool => {
-			// 0. Check if the tool was disabled via the tool picker. If so, it must be disabled here
-			const toolPickerSelection = request.tools.get(getContributedToolName(tool.name));
-			if (toolPickerSelection === false) {
-				return false;
-			}
-
-			// 1. Check for what the consumer wants explicitly
-			const explicit = filter?.(tool);
-			if (explicit !== undefined) {
-				return explicit;
-			}
-
-			// 2. Check if the request's tools explicitly asked for this tool to be enabled
-			for (const ref of request.toolReferences) {
-				const usedTool = toolMap.get(ref.name);
-				if (usedTool?.tags.includes(`enable_other_tool_${tool.name}`)) {
-					return true;
+		return this.tools
+			.map(tool => {
+				// Apply model-specific alternative if available via alternativeDefinition
+				const owned = this._copilotTools.value.get(getToolName(tool.name) as ToolName);
+				if (owned?.alternativeDefinition) {
+					const alternative = owned.alternativeDefinition(endpoint);
```

Comment:
We call alternativeDefinition in the tools getter, would it make sense to change that to be model-specific? Like `getTools(model)`? Or should it stay generic and we only apply the model-specific stuff here?